autocomplete.js

```diff
@@ -0,0 +1,122 @@
+const TestrailApiClient = require("testrail-api");
+const joinUrlParts = require("url-join");
+const { default: axios } = require("axios");
+
+const { sanitizeHostname } = require("./helpers");
+
+function createAutocompleteFunction({
+  fetchResult,
+  itemsDataPath,
+  idPath,
+  valuePath,
+  checkForNestedItems,
+}) {
+  return async (query, params) => {
+    const {
+      hostname,
+      username,
+      apiKey,
+    } = params;
+
+    const testRailClient = new TestrailApiClient({
+      host: sanitizeHostname(hostname),
+      user: username,
+      password: apiKey,
+    });
+
+    const { body: result } = await fetchResult(testRailClient, params);
+    const items = itemsDataPath ? result[itemsDataPath] : result;
+
+    if (checkForNestedItems) {
+      const nestedItems = items.reduce((acc, cur) => {
+        if (!cur[itemsDataPath]) {
+          return acc;
+        }
+        return [...acc, ...cur[itemsDataPath]];
+      }, []);
+      items.push(...nestedItems);
+    }
+
+    return mapAndFilterItems(items, query, { valuePath, idPath });
+  };
+}
+
+async function listUsers(query, params) {
+  const {
+    projectId,
+    hostname,
+    username,
+    apiKey,
+  } = params;
+  const url = joinUrlParts(
+    sanitizeHostname(hostname),
+    "index.php?/api/v2/get_users",
+    String(projectId),
+  );
+
+  const { data } = await axios({
+    method: "GET",
+    url,
+    auth: {
+      username,
+      password: apiKey,
+    },
+  });
+
+  return mapAndFilterItems(data, query, {});
+}
+
+function mapAndFilterItems(
+  items,
+  query,
+  {
+    valuePath = "name",
+    idPath = "id",
+  },
+) {
+  const autocompleteItems = items.map((item) => ({
+    id: String(item[idPath]),
+    value: item[valuePath],
+  }));
+
+  if (!query) {
+    return autocompleteItems;
+  }
+
+  const lowerCaseQuery = query.toLowerCase();
+  return autocompleteItems.filter((item) => (
+    item.id.toLowerCase().includes(lowerCaseQuery)
+    || item.value.toLowerCase().includes(lowerCaseQuery)
+  ));
+}
+
+module.exports = {
+  listProjects: createAutocompleteFunction({
+    fetchResult: (testRailClient) => testRailClient.getProjects(),
```

Comment:
That's correct.

Reply:
Since we're responsible for processing the information obtained from their API, we should convert it to the desired format upon reception, not right before using them.

Reply:
The autocomplete's item's id has to be of type string, we can't really use the numbers here. The plugin method's receives values as string and parses it to number, which is the format required by the TestRail package.